go/teams/list.go

```diff
@@ -45,6 +45,113 @@ func getTeamsListFromServer(ctx context.Context, g *libkb.GlobalContext, uid key
 	return list.Teams, nil
 }
 
+func memberNeedAdmin(member keybase1.MemberInfo, meUID keybase1.UID) bool {
+	return member.UserID == meUID &&
+		(member.Role.IsAdminOrAbove() || (member.Implicit != nil && member.Implicit.Role.IsAdminOrAbove()))
+}
+
+// verifyMemberRoleInTeam checks if role give in MemberInfo matches
+// what team chain says. Nothing is checked when MemberInfo's role is
+// NONE, in this context it means that user has implied membership in
+// the team and no role given in sigchain.
+func verifyMemberRoleInTeam(ctx context.Context, member keybase1.MemberInfo, team *Team) error {
+	if member.Role == keybase1.TeamRole_NONE {
+		return nil
+	}
+
+	memberUV, err := team.chain().GetLatestUVWithUID(member.UserID)
+	if err != nil {
+		return err
+	}
+	role, err := team.chain().GetUserRole(memberUV)
+	if err != nil {
+		return err
+	}
+	if role != member.Role {
+		return fmt.Errorf("unexpected member role: got %v but actual role is %v", member.Role, role)
+	}
+	return nil
+}
+
+// getTeamForMember tries to load team in a recent enough state to
+// contain member with correct role as set in MemberInfo. It might
+// trigger a reload with ForceRepoll if cached state does not match.
+func getTeamForMember(ctx context.Context, g *libkb.GlobalContext, member keybase1.MemberInfo, needAdmin bool) (*Team, error) {
+	team, err := Load(ctx, g, keybase1.LoadTeamArg{
+		ID:          member.TeamID,
+		NeedAdmin:   needAdmin,
```

Comment:
throw in `Public: member.TeamID.IsPublic()` if this should work on public teams. Sorry that's weird. (same on below Load)